src/lua/init.lua

```diff
@@ -86,77 +114,85 @@ local function load_lua(file)
     return loadfile(file)
 end
 
-local function search_cwd_lib(name)
-    local path = "./?."..soext
-    return package.searchpath(name, path)
-end
+local function merge_strings(paths, templates)
+    local searchpaths = {}
+    paths = type(paths) == 'table' and paths or { paths }
+    templates = templates or {}
 
-local function search_cwd_lua(name)
-    local path = "./?.lua;./?/init.lua"
-    return package.searchpath(name, path)
-end
+    if #templates == 0 then
+        return paths
+    end
 
-local function traverse_rocks(name, pathes_search)
-    local cwd = fio.cwd()
-    local index = string.len(cwd) + 1
-    local strerr = ""
-    while index ~= nil do
-        cwd = string.sub(cwd, 1, index - 1)
-        for i, path in ipairs(pathes_search) do
-            local file, err = package.searchpath(name, cwd .. path)
-            if err == nil then
-                return file
-            end
-            strerr = strerr .. err
+    if #paths == 0 then
+        return templates
+    end
+
+    for _, path in ipairs(paths) do
+        for _, template in pairs(templates) do
+            table.insert(searchpaths, fio.pathjoin(path, template))
         end
-        index = string.find(cwd, "/[^/]*$")
     end
-    return nil, strerr
+
+    return searchpaths
 end
 
-local function search_rocks_lua(name)
-    local pathes_search = {
-        "/.rocks/share/tarantool/?.lua;",
-        "/.rocks/share/tarantool/?/init.lua;",
-    }
-    return traverse_rocks(name, pathes_search)
+local LIB_TEMPLATES = { '?.'..soext }
+local LUA_TEMPLATES = { '?.lua', '?/init.lua' }
+local ROCKS_LIB_TEMPLATES = { ROCKS_LIB_PATH .. '/?.'..soext }
+local ROCKS_LUA_TEMPLATES = { ROCKS_LUA_PATH .. '/?.lua', ROCKS_LUA_PATH .. '/?/init.lua' }
+
+local function search_lib(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LIB_TEMPLATES), ';'))
 end
 
-local function search_rocks_lib(name)
-    local pathes_search = {
-        "/.rocks/lib/tarantool/?."..soext
-    }
-    return traverse_rocks(name, pathes_search)
+local function search_lua(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LUA_TEMPLATES), ';'))
 end
 
-local function cwd_loader_func(lib)
-    local search_cwd = lib and search_cwd_lib or search_cwd_lua
-    local load_func = lib and load_lib or load_lua
+local function traverse_path(path)
+    path = path:sub(1, 1) == '/' and path or fio.abspath(path)
```

Comment:
```console
> traverse_path('/a/b/c/../../A')
---
- - /a/b/c/../../A
  - /a/b/c/../..
  - /a/b/c/..
  - /a/b/c
  - /a/b
  - /a
  - /
...
```

Is this appropriate behavior? Maybe `fio.abspath` should be called always?

Reply:
This is how it works right now. Your suggestion is reasonable and I agree with it, but it is a changing behaviour.

Reply:
Ok, it's reasonable to keep it then.